### PR TITLE
GRF-231  internal API endpoint to GET a category for edition purpose

### PR DESCRIPTION
**Pull request by FSAkeneo**
**Description (for Contributor and Core Developer)**

New endpoint /category/rest/<id> to get information about categories.
contains
- core category data : id, parent_id, code,labels
- permissions settings
- attributes values

Values are hardcoded and return attribute values corresponding to the (yet to exist on backend) hardcoded template.

The Query Bus has not been implemented, the http controller simply build the query and pass it directly to the right handler.
The handler builds the response and returns it.
Note the conversion of models between domain and API.
models in API are destined to be published amongst all PIM devs, models in the domain can evolve freely.

### CODE REVIEW

**Comment by BryanS29-akeneo** [comment]
Why re-create models in `back\Api` directory ? Can't we use the ones that are in `back/Domain`?
Having same models is a bit confusing

**Comment by etiennekerbastard** [comment]
> 

Yep if they're identical we can use the one in `back\Domain` imo

File: src/Akeneo/Category/back/Application/GetCategoryHandler.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                'data' => 'Shoes Slippers Sneakers',
+                'locale' => 'en_US',
+            ],
+            'seo_keywords_54f6725a-17e1-11ed-a002-73412755f3bd_fr_FR' => [
+                'data' => 'Chaussures Tongues Espadrilles',
+                'locale' => 'fr_FR',
+            ],
+        ]);
+
+        $readModel = new CategoryReadModel(
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Application/GetCategoryHandler.php:82]: Move Category model in controller for the serialization

File: src/Akeneo/Category/back/Api/Model/CategoryReadModel.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    public function permissions(): Permissions
+    {
+        return $this->permissions;
+    }
+
+    public function attributeValues(): AttributeValues
+    {
+        return $this->attributeValues;
+    }
+
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Api/Model/CategoryReadModel.php:39]: add method fromDomain() => retrieve a data from database.

File: src/Akeneo/Category/back/Application/GetCategoryHandler.php (hunk 1 above)
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Application/GetCategoryHandler.php:82]: Rename class to category and move to Api/model/read

File: src/Akeneo/Category/back/Application/GetCategoryHandler.php [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            Category::fromDomainModel($domainCategory),
+            $permissions,
+            $attributeValues,
+        );
+        // returning the category as seen by the outside (whoever created and dispatched the query)
+        return $readModel;
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Application/GetCategoryHandler.php:88]: return read model category build with Category model (check permission for edit page)

File: src/Akeneo/Category/back/Application/GetCategoryHandler.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            'fr_FR' => 'Vêtements',
+            'en_US' => 'Clothes',
+            'de_DE' => 'Kleidung',
+        ]);
+
+        // the category as used internally
+        $domainCategory = new CategoryFromDomain($query->categoryId(), $code, $labels, new CategoryId(1));
+
+        /* @var $permission Permissions */
+        $permissions = new Permissions([
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Application/GetCategoryHandler.php:38]: check ACL pimee_enrich_category_edit_permissions and feature flag permission. (if no right => ALL ou [] = no permission. To check)

File: src/Akeneo/Category/back/Domain/ValueObject/CategoryId.php [hunk 5]
```diff
@@ -21,4 +21,9 @@ public function getId(): int
     {
         return $this->id;
     }
+
+    public function __toString(): string
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Domain/ValueObject/CategoryId.php:25]: no need

File: src/Akeneo/Category/back/Infrastructure/Symfony/Resources/config/routing/internal_api/category.yml [hunk 6]
```diff
@@ -1,6 +1,15 @@
...
   defaults:
     _controller: Akeneo\Category\back\Infrastructure\Controller\InternalAPI\UpdateCategoryController
     _feature: enriched_category
   methods: [POST]
+
+pim_category_rest_get:
+  path: /rest/{id}
+  requirements: { 'id': '\d+' }
+  defaults:
+    _controller: Akeneo\Category\back\Infrastructure\Controller\InternalAPI\GetCategoryController
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Infrastructure/Symfony/Resources/config/routing/internal_api/category.yml:13]: check path internalApi directory and no back

File: src/Akeneo/Category/back/tests/.php_cs.php [hunk 7]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+return (new PhpCsFixer\Config())
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/tests/.php_cs.php:3]: desactivate Yoda style

File: src/Akeneo/Category/back/Api/Model/Category/Category.php [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        return new Category(
+            CategoryId::fromDomainModel($c->getId()),
+            Code::fromDomainModel($c->getCode()),
+            LabelCollection::fromDomainModel($c->getLabelCollection()),
+            CategoryId::fromDomainModel($c->getParentId()),
+        );
+    }
+
+    public function __construct(
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Api/Model/Category/Category.php:28]: private

File: src/Akeneo/Category/back/Api/Model/Category/Category.php [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+use Akeneo\Category\Domain\Model\Category as CategoryFromDomain;
+
+/**
+ * This model represents the core information about a category as exposed to the outside of the category bounded context
+ * It resembles the eponymous internal domain model but can drift in the future.
+ *
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class Category
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Api/Model/Category/Category.php:16]: To delete, see ReadModel upgrade.

File: src/Akeneo/Category/back/Api/Model/Category/CategoryId.php [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+use Webmozart\Assert\Assert;
+
+/**
+ * This model represents a category ID as exposed to the outside of the category bounded context
+ * It resembles the eponymous internal domain model but can drift in the future.
+ *
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class CategoryId
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Api/Model/Category/CategoryId.php:17]: Use primitif int instead

File: src/Akeneo/Category/back/Api/Model/Category/Code.php [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+use Webmozart\Assert\Assert;
+
+/**
+ * This model represents a category code as exposed to the outside of the category bounded context
+ * It resembles the eponymous internal domain model but can drift in the future.
+ *
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class Code
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Api/Model/Category/Code.php:17]: use string instead

File: src/Akeneo/Category/back/Api/Model/Category/LabelCollection.php [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * This model represents labels of a category code as exposed to the outside of the category bounded context
+ * It resembles the eponymous internal domain model but can drift in the future.
+ *
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ * @phpstan-type Locale string
+ * @phpstan-type LocalizedLabels array<Locale, string>
+ */
+final class LabelCollection
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Api/Model/Category/LabelCollection.php:17]: use array instead

File: src/Akeneo/Category/back/Api/Model/CategoryReadModel.php [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+namespace Akeneo\Category\Api\Model;
+
+use Akeneo\Category\Api\Model\Category\Category;
+
+class CategoryReadModel
+{
+    public function __construct(
+        private Category $category,
```
Comment (antoinetrouve) [inline src/Akeneo/Category/back/Api/Model/CategoryReadModel.php:19]: Use only category model

**Review by antoinetrouve — CHANGES_REQUESTED** [review]

File: src/Akeneo/Category/back/tests/.php_cs.php (hunk 7 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/tests/.php_cs.php:3]: done

File: src/Akeneo/Category/back/Infrastructure/Symfony/Resources/config/routing/internal_api/category.yml (hunk 6 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Infrastructure/Symfony/Resources/config/routing/internal_api/category.yml:13]: Done

File: src/Akeneo/Category/back/Application/GetCategoryHandler.php [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * a GetCategoryHandler executes GetCategoryQuery queries.
+ *
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class GetCategoryHandler
+{
+    public function __invoke(GetCategoryQuery $query): CategoryReadModel
```
[the remark this answers is not in the document]
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Application/GetCategoryHandler.php:25]: done

File: src/Akeneo/Category/back/Application/GetCategoryHandler.php (hunk 1 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Application/GetCategoryHandler.php]: Done

File: src/Akeneo/Category/back/Api/Model/CategoryReadModel.php (hunk 2 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Api/Model/CategoryReadModel.php]: Done

File: src/Akeneo/Category/back/Api/Model/Category/Category.php (hunk 9 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Api/Model/Category/Category.php]: Done

File: src/Akeneo/Category/back/Api/Model/Category/Code.php (hunk 11 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Api/Model/Category/Code.php]: Done

File: src/Akeneo/Category/back/Api/Model/CategoryReadModel.php (hunk 13 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Api/Model/CategoryReadModel.php]: Done

File: src/Akeneo/Category/back/Application/GetCategoryHandler.php (hunk 1 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Application/GetCategoryHandler.php]: Done

File: src/Akeneo/Category/back/Api/Model/Category/CategoryId.php (hunk 10 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Api/Model/Category/CategoryId.php]: Done

File: src/Akeneo/Category/back/Api/Model/Category/LabelCollection.php (hunk 12 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Api/Model/Category/LabelCollection.php]: Done

**Review by antoinetrouve — APPROVED** [review]

**Review by BryanS29-akeneo — APPROVED** [review]

File: src/Akeneo/Category/back/Application/GetCategoryQueryHandler.php [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+        $labels = [
+            'fr_FR' => 'Vêtements',
+            'en_US' => 'Clothes',
+            'de_DE' => 'Kleidung',
+        ];
+
+        // the category as used internally
+//        $domainCategory = new Category($query->categoryId(), $code, $labels, new CategoryId(1));
+
+        // TODO : check ACL and Feature flag (see PR review)
```
Comment (amigaire) [inline src/Akeneo/Category/back/Application/GetCategoryQueryHandler.php:36]: Has it been viewed?

**Comment by olivierpontier** [comment]
Are the models `src/Akeneo/Category/back/Api/Model/Permissions.php` and `src/Akeneo/Category/back/Api/Model/AttributeValues.php` still useful? Did we agree on exposing only the `Model\Read\Category`?

**Comment by antoinetrouve** [comment]
> Are the models `src/Akeneo/Category/back/Api/Model/Permissions.php` and `src/Akeneo/Category/back/Api/Model/AttributeValues.php` still useful? Did we agree on exposing only the `Model\Read\Category`?

Yes you are right. I just foget to delete it.

File: src/Akeneo/Category/back/Application/GetCategoryQueryHandler.php (hunk 15 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Application/GetCategoryQueryHandler.php]: Yes it still to do when we 'll work on the permission.

**Review by etiennekerbastard — APPROVED** [review]

**Pull request merged**